Add exponential backoff for erroring jobs

sagelywizard opened the pull request:
This PR adds an exponential backoff for rescheduling erroring jobs. It does this by counting the number of consecutive times which a job has errored in its recorded history (up to a maximum) and using that to calculate the amount of time to wait before rescheduling the job.

This PR also adds a config parameter which is important for exponential backoff: `max_history`. This sets the maximum number of `started`, `stopped`, and `crashed` events which the job will track. Setting a large history will allow the job to be suspended for longer (up to a configurable maximum) but will cause higher memory usage.

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -259,25 +267,57 @@ remove_job_int(#job{} = Job) ->
+    Now = unix_now(),
+    % Select jobs where the pid is undefined and the exponential backoff
+    % threshold has been crossed.
+    MatchSpec = [{#job{pid = '$1',errored_at = '$2',errors = '$3', _='_'},
```
Should we use fun2ms here, for clarity? It was a bit tricky deciphering the meaning of this expression

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -259,25 +267,57 @@ remove_job_int(#job{} = Job) ->
+    % Select jobs where the pid is undefined and the exponential backoff
+    % threshold has been crossed.
+    MatchSpec = [{#job{pid = '$1',errored_at = '$2',errors = '$3', _='_'},
+        [{'not',{is_pid,'$1'}},
```
Maybe in the head just say `#job{pid = undefined}` to save an extra guard. Should probably also update typespec of job records' pid field to be `pid() | undefined`

sagelywizard replied:
Good thought.

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -259,25 +267,57 @@ remove_job_int(#job{} = Job) ->
 -spec running_job_count() -> non_neg_integer().
 running_job_count() ->
-    ets:info(?MODULE, size) - pending_job_count().
+    MatchSpec = [{#job{pid = '$1', _='_'},
```
Would fun2ms or qlc here be too hard on performance?

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -354,3 +394,9 @@ update_history(Job, Type, When) ->
+-spec unix_now() -> float().
+unix_now() ->
+    {Meg, Sec, Mic} = os:timestamp(),
+    Meg*1000000+Sec+(Mic/1000000).
```
Minor nit: spacing between operators and operands. For consistency.

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -30,23 +30,25 @@
 -define(DEFAULT_MAX_JOBS, 100).
 -define(DEFAULT_MAX_CHURN, 20).
 -define(DEFAULT_SCHEDULER_INTERVAL, 60000).
+-define(ERROR_BACKOFF_MILLIS, 30000).
```
Saw below we calculate back-off interval in seconds and then translate to milliseconds to compare with this constant.  Maybe just use seconds here? ... Or would that clash visually with `DEFAULT_SCHEDULER_INTERVAL` which is in milliseconds...

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -30,23 +30,25 @@
-          history :: history()}).
+          history :: history(),
+          errors :: integer(),
+          errored_at :: integer()}).
```
Is this an integer or a float? Initialized below to 0.0 but not sure if it we still mean to be float.

sagelywizard replied:
Should've been `float`.

nickva commented:
Should we squash the commits in the pr?  Most are to the same file or same code blocks. Fixes during review could be separate commits if we want to track that.

nickva commented:
For exponential back-off I think `uniform_random(2^ConsecutiveErrorCount)` might be better. TCP uses something like that. 

The difference being it avoids global synchronization between jobs. Say jobs replicate from one source, it fails, then all jobs will start retrying at the same schedule.

nickva commented:
To avoid the complicated multi-line matchspecs, was wondering if perhaps it is better to use a `next_try_at` field in the job record?

When job stops, depending how it stops, we schedule it then when to run next by setting `next_try_at = uniform_random(2^ConsecutiveErrorCount * ?ERROR_BACKOFF)`. 

Then we don't have to to compute all that stuff during match spec and it just compares next_try_at with Now.

sagelywizard commented:
For the `try_next_at` field idea, I had thought of that but it makes it difficult to change the delay interval configuration dynamically. I dunno if that's a big deal, but it's why I went for the matchspec approach.

sagelywizard commented:
Yeah, I'll definitely squash this PR down once I get a "+1 pending squash".

sagelywizard commented:
Good point about the random uniform thing. Though do we need it since we have the max churn stuff? It'd require a `try_next_at` field and I'm not sure we want to do that. @rnewson @nickva Thoughts?

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -207,14 +207,17 @@ not_recently_crashed(#job{} = Job) ->
             true;
-        [{crashed, When} | _] ->
+        [{{crashed, _Reason}, When} | _] = Crashes ->
+            BackoffExp = erlang:min(1 bsl length(Crashes), ?MAX_BACKOFF_EXPONENT),
```
`?MAX_BACKOFF_EXPONENT` is not the exponent though, `length(Crashes)` is the exponent.

sagelywizard replied:
Hrm, `?MAX_BACKOFF_EXPONENT` may be a bit of a misnomer, but the code looks correct to me. Could be `1 bsl erlang:min(length(Crashes), ?MAX_BACKOFF_EXPONENT)`. I'll make that change.

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -266,57 +260,24 @@ remove_job_int(#job{} = Job) ->
-        0}],
-        [true]}
-    ],
+    MatchSpec = [{#job{pid='$1', _='_'}, [{'not', {'is_pid', '$1'}}], [true]}],
```
Should this  be `#job{pid=undefined, ...}` ? 

Going by idea that pending jobs are the ones which are not running but a ready to run, so they don't have a pid yet.

sagelywizard replied:
Changed

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -266,57 +260,24 @@ remove_job_int(#job{} = Job) ->
-        ['$_']}
-    ],
-    ets:select(?MODULE, MatchSpec).
+    ets:tab2list(?MODULE) -- pending_jobs().
```
Superficially looking at it, this could be costly. First we'd make a full copy of all the jobs in tab2list. Then a copy in pending_jobs(). Then another copy is the resulting list. L1--L2 complexity is about O(length(L1)*length(L2)) as well.

Since we are returning whole objects perhaps `ets:match_object(Tab, Pattern)` would work here.

sagelywizard replied:
Yeah, it is rather costly. I don't think you could use `ets:match_object` since we want to filter by type. The matchspec/select which I reverted is less costly, so I could un-revert that.

sagelywizard commented:
That was a rebase onto 63012-scheduler

nickva commented on src/couch_replicator_scheduler.erl:
```diff
@@ -225,17 +232,22 @@ oldest_job_first(#job{} = A, #job{} = B) ->
     end.
 
 
 crash_history(#job{} = Job) ->
```
I think we want `JobHistoryEvents :: list()` here as arg. Or maybe it is cleaner to push EventsBeforeStop processing to `crash_history/ 1` to make not_recently_crashed a bit smaller visually.

rnewson replied:
I think crash_history should return the crash history, we've mingled in a separate concern here.

nickva replied:
Agree. Was thinking with my 10pm brain there trying to get tests to pass. If crash history is not called from anywhere else, the fix could rename the function, as well...

nickva commented:
+1

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -148,8 +149,14 @@ handle_info({'DOWN', _Ref, process, Pid, Reason}, State) ->
-            true = ets:insert(?MODULE, Job1),
+            Job1 = Job0#job{pid = undefined},
+            Job2 = case Reason of
+                normal ->
```
given the previous function clause, how could Reason be normal here?

sagelywizard replied:
Whoops, I'll fix that.

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -30,13 +30,14 @@
 
 %% types
--type event_type() :: started | stopped | crashed.
+-type event_type() :: started | stopped | {crashed, any()}.
```
where is the Reason term used? since it'll often be large, and we'll have lots of jobs in memory I think we might not want to store it.

sagelywizard replied:
For 1) introspection and 2) the ability to make scheduling decisions based on specific errors. Thoughts?

rnewson replied:
ok

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -284,13 +296,12 @@ running_job_count() ->
 -spec running_jobs() -> [#job{}].
 running_jobs() ->
-    ets:tab2list(?MODULE) -- pending_jobs().
+    ets:select(?MODULE, [{#job{pid = '$1', _='_'}, [{is_pid, '$1'}], ['$_']}]).
```
nice. can we pull this into a separate commit, though? (since it's not about adding exponential backoff)

sagelywizard replied:
Sure

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -226,16 +233,21 @@ oldest_job_first(#job{} = A, #job{} = B) ->
-            true;
-        [{crashed, When} | _] ->
+        [{{crashed, _Reason}, When} | _] = Crashes ->
+            BackoffExp = erlang:min(length(Crashes), ?MAX_BACKOFF_EXPONENT),
```
off by one here. the first time we back off will be for _two_ backoff intervals not one. suggest length(Crashes) - 1.

nickva replied:
Or make backoff =  random:uniform(interval)? The expected back off will be one interval.

sagelywizard replied:
I think random would make sense if the backoff exponent was persisted in memory. If the backoff was randomly generation for each filtration of crashed jobs, jobs will be biased toward low exponents.

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -226,16 +233,21 @@ oldest_job_first(#job{} = A, #job{} = B) ->
+        [{{crashed, _Reason}, When} | _] = Crashes ->
+            BackoffExp = erlang:min(length(Crashes), ?MAX_BACKOFF_EXPONENT),
+            BackoffInterval = (1 bsl BackoffExp) * ?BACKOFF_INTERVAL_MICROS,
+            couch_log:notice("delay for ~p is now ~p s", [Job#job.id, BackoffInterval div 1000000]),
```
remove log line, this'll just be noise. Perhaps move the logic here to a separate side-effect free function instead, so we can write some eunit tests to prove the progression of timeouts (and their cap).

sagelywizard replied:
K

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -226,16 +233,21 @@ oldest_job_first(#job{} = A, #job{} = B) ->
 
 not_recently_crashed(#job{} = Job) ->
     case crash_history(Job) of
```
if the most recent history item is a `stopped`, we could return true here. That is, we can define that a job that last ran successfully has not recently crashed.

sagelywizard replied:
Sure.

sagelywizard committed:
Make max_history configurable

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -129,28 +142,23 @@ handle_cast(_, State) ->
-        _Else ->
-            ok
-    end,
+handle_info({'DOWN', _Ref, process, _Pid, normal}, State) ->
```
why remove the job cleanup when a replication completes?

nickva replied:
Check latest `63012-scheduler` branch (rebase perhaps), there were changes  related to monitoring there.

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -205,52 +213,62 @@ handle_config_terminate(Self, _, _) ->
     Runnable2 = lists:filter(fun not_recently_crashed/1, Runnable1),
     Runnable3 = lists:sublist(Runnable2, Count),
-    lists:foreach(fun start_job_int/1, Runnable3).
+    lists:foreach(fun(J) -> start_job_int(J, MaxHistory) end, Runnable3).
```
that's a real shame, breaks hot upgrade. can't think of a way around it right now but :(

nickva replied:
A list comprehension could do it then. Just return ok, in the end to stay consistent with the existing return type

sagelywizard replied:
Yeah, a LC would work. I'll change that.

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -317,38 +334,38 @@ job_by_id(Id) ->
 
-start_pending_jobs(Max) ->
-    start_pending_jobs(Max, running_job_count(), pending_job_count()).
+start_pending_jobs(Max, MaxHistory) ->
```
'Max' is now ambiguous, 'MaxJobs'?

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -317,38 +334,38 @@ job_by_id(Id) ->
-rotate_jobs(MaxJobs, MaxChurn, Running, Pending) when Running == MaxJobs, Pending > 0 ->
-    stop_jobs(min([Pending, Running, MaxChurn])),
-    start_jobs(min([Pending, Running, MaxChurn]));
+rotate_jobs(MaxJobs, MaxChurn, Running, Pending, MaxHistory) when Running == MaxJobs, Pending > 0 ->
```
add it after MaxChurn, that's just untidy.

rnewson commented on src/couch_replicator_scheduler_job.erl:
```diff
@@ -450,11 +450,8 @@ format_status(_Opt, [_PDict, State]) ->
-    % should be
-    ok.
+report_job_error(Rep, Error) ->
+    exit({error, Rep, Error}).
```
this seems to have no bearing on the change, what's this for?

sagelywizard replied:
It'll cause the job to terminate in a non-successful way, which is necessary for exponential backoff.

sagelywizard replied:
Though I'm a bit on the fence about passing the `Rep` state back, as it can be pretty big. The reason for the error is useful/necessary though.

nickva replied:
Tested locally. Killed a replication by hand noticed that keeping `#rep{}` record in each crash event could be a problem. I leaks credentials to logs and is a rather large chunk of data (compared to other things there).

Also not sure if we should exit inside the context of the terminate/2 callback. It is a bit unusual. This was a place where we wrote errors back to documents. But if we don't plan on keeping #rep{} process will already exit with `Reason` so we can just not do anything. In a separate pr can remove report_job_error altogether.

rnewson replied:
it's unnecessary. killing a running replication job extends the history with a crash event without this change. remove report_job_error entirely, it does nothing.

rnewson commented:
as a meta comment, I think it might be time we pass State into most of the private functions. we're now passing three bits of it down, in addition to other parameters.

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -205,52 +213,62 @@ handle_config_terminate(Self, _, _) ->
-        [{crashed, When} | _] ->
-            timer:now_diff(os:timestamp(), When)
-                >= ?MINIMUM_CRASH_INTERVAL
+        [stopped | _] ->
```
`{stopped, _When}`, right?

sagelywizard replied:
Whoops

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -205,52 +213,62 @@ handle_config_terminate(Self, _, _) ->
+            Crashes = lists:filter(
+                fun({{crashed, _Reason}, _When}) -> true; (_) -> false end,
+                EventsBeforeStop
+            ),
```
why lists:filter instead of the simpler and shorter list comprehension?

sagelywizard replied:
The comprehension was getting a bit long, and I prefer multiline funs rather than multiline list comprehensions, but I could change it back to a long list comprehension.

nickva commented:
There is a conflict with 63012-scheduler in the `handle_info({'DOWN', ...)` function clause

nickva commented:
Unit tests pass on this branch

```
=======================================================
  All 171 tests passed.
```

Did manual testing. Exponential back-off seems to work (manually removed a target db) and so job staying in the crashed state for a while even after adding it back.

A few more comments above though, about `exit({error, ..})`

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -317,38 +344,38 @@ job_by_id(Id) ->
-    start_pending_jobs(MaxJobs, Running, Pending),
-    rotate_jobs(MaxJobs, MaxChurn, Running, Pending).
+    stop_excess_jobs(MaxJobs, Running, MaxHistory),
+    start_pending_jobs(MaxJobs, MaxHistory, Running, Pending),
```
inconsistent parameter ordering.

rnewson commented:
needs an extra clause in handle_config_change to cast to set_max_history

sagelywizard commented:
Yet another rebase.

sagelywizard committed:
Pass state through scheduler rather than configs

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -320,40 +345,44 @@ job_by_id(Id) ->
 
+rotate_jobs(State, Running, Pending) ->
+    #state{max_jobs=MaxJobs, max_history=MaxHistory, max_churn=MaxChurn} = State,
+    if Running == MaxJobs, Pending > 0 ->
```
extract MaxJobs in the function head so the guard clause can work, and extract the others in that `= State` line.

rnewson commented:
+1

sagelywizard committed:
Add exponential backoff for crashing jobs

This commit adds an exponential backoff on retry for crashing jobs. It
does this by looking at the event history for a job and filtering jobs
which have crashed multiple times in a row.

Note that the exponential backoff is bound by two things:

- The `MAX_BACKOFF_EXPONENT` config parameter, and
- The `MAX_HISTORY` config parameter

Only `MAX_HISTORY / 2` `crashed` events can be recorded. This limits the
potential size of the backoff exponent.

sagelywizard committed:
Improve efficiency of ets queries

sagelywizard commented:
That was the squash/rebase.
